apps/docs/pages/guides/auth/auth-postgres-hooks.mdx

```diff
@@ -0,0 +1,614 @@
+  returns void language plpgsql as $$
+begin
+  insert into test_results(name, result, debug) values (name, result, debug);
+end;
+$$;
+
+do $$
+  declare
+     -- Arrange and setup any variables used in the test
+    _user1 uuid = '00000000-0000-0000-0000-000000000001';
+    _factor1 uuid = 'FFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFF1';
+    _valid_user1_event jsonb = jsonb_build_object('user_id', _user1, 'factor_id', _factor1, 'valid', true);
+    _invalid_user1_event jsonb = jsonb_build_object('user_id', _user1, 'factor_id', _factor1, 'valid', false);
+    _result jsonb;
+  begin
+    set search_path to 'public';
+
+    insert into auth.users(id) values (_user1);
+
+    -- Act and run the function
+    perform tests.record_result(
+      'Valid event result'
+      , (select hook_mfa_verification_attempt(_valid_user1_event)) ->> 'decision' = 'continue'
+    );
+
+    -- Add further test cases here
+  end;
+$$;
+
+set search_path to 'extensions';
+
+-- Assert that the desired result is received
+select set_eq(
+  $$ select name, result, debug from test_results $$,
+  $$ select * from (values
+    , ('Valid event result', true, null)
+  ) as t $$,
+  'MFA Auth hook tests:'
+);
+
+select * from finish();
+rollback;
+````
+
+
+### Connecting the Hook
+
+In the dashboard, navigate to `Authentication > Hooks (Beta)` and select the appropriate PostgreSQL function from the dropdown menu.
+![Hooks UI](/docs/img/guides/auth-postgres-hooks/hooks-ui.png)
+
+
+## MFA Verification Attempt Hook
+
+You can add additional checks to the [Supabase MFA implementation](https://supabase.com/docs/guides/auth/auth-mfa) with hooks. For example, you can:
+
+- Limit the number of verification attempts performed over a period of time.
+- Sign out users who have too many invalid verification attempts.
+- Count, rate limit, or ban sign-ins.
+
+### Inputs
+
+Supabase Auth will supply the following fields to your hook:
+
+- `factor_id` Unique identifier for the MFA factor being verified. 
+- `user_id` Unique identifier for the user.
+- `valid` Whether the verification attempt was valid. For TOTP, this means that the six digit code was correct (true) or incorrect (false).
+
+Example payload:
+
+```json
+{
+	"factor_id": "6eab6a69-7766-48bf-95d8-bd8f606894db",
+	"user_id": "3919cb6e-4215-4478-a960-6d3454326cec",
+	"valid": true,
+}
+```
+
+### Outputs
+
+Return this if your hook processed the input without errors.
+
+- `decision` A string containing the decision on whether to allow authentication to move forward. Use `reject` to deny the verification attempt and log the user out of all active sessions. Use `continue` to use the default Supabase Auth behavior.
+- `message` The message to show the user if the decision was `reject`.
+
+Example output:
+
+```json
+{
+   "decision": "reject",
+   "message" : "You have exceeded maximum number of MFA attempts."
+}
+```
+### Example: Limit Failed MFA Verification Attempts
+<Tabs
```

Comment:
Right now the tabs look like this so there's not much purpose to having the tab:

<img width="883" alt="image" src="https://github.com/supabase/supabase/assets/26616127/29924ea8-ce91-421b-acc4-ecb6c6a9bdeb">

I think what you want to do is have one global `Tabs` section, then have three `TabPanels` within it corresponding to the three different hooks

Reply:
That could work too - I've kept the tabs gone ahead and added a draft example for each extensibility point but let me know if another style of organizing the content is preferred